### PR TITLE
2.x Adding interest channel eviction mechanism + tests.

### PR DESCRIPTION
**Pull request by qiangdavidliu**

### CODE REVIEW

File: eureka2-client/src/main/java/com/netflix/eureka2/client/channel/InterestChannelInvoker.java [hunk 1]
```diff
@@ -28,6 +32,11 @@ public InterestChannelInvoker(InterestChannel delegate, EurekaClientMetricFactor
     public InterestChannelInvoker(InterestChannel delegate, EurekaClientMetricFactory metricFactory, Scheduler scheduler) {
         super(metricFactory.getSerializedTaskInvokerMetrics(InterestChannelInvoker.class), scheduler);
         this.delegate = delegate;
+        if (delegate instanceof Sourced) {
```
Comment (tbak) [inline eureka2-client/src/main/java/com/netflix/eureka2/client/channel/InterestChannelInvoker.java]: Shouldn't we have always single case here with sourced delegate?

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    private SourcedEurekaRegistryImpl registry;
+    private RetryableConnection<InterestChannel> retryableConnection;
+    private ReplaySubject<InterestChannel> channelSubject;
+
+    private MyInterestClient client;
+
+    @Before
+    public void setUp() {
+        registry = mock(SourcedEurekaRegistryImpl.class);
```
Comment (tbak) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: Why not mock SourcedEurekaRegistry which is an interface?

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        channelSubject.onNext(channel2);
+        channel1Lifecycle.onError(new Exception("test error"));
+
+        channel2BatchSubject.onNext(newBufferStart());
+        verify(registry, never()).evictAll(any(Source.SourceMatcher.class));
+
+        channel2BatchSubject.onNext(newBufferEnd());
+        verify(registry, times(1)).evictAll(any(Source.SourceMatcher.class));
+        Thread.sleep(1000);
```
Comment (tbak) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: Why this sleep here?

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            return retryableConnection;
+        }
+
+        @Override
+        public Observable<ChangeNotification<InstanceInfo>> forInterest(Interest<InstanceInfo> interest) {
+            return null;
+        }
+
+        @Override
+        public void registryEvictionSubscribe(RetryableConnection<InterestChannel> retryableConnection) {
```
Comment (tbak) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: Why override here if we only call super?

File: eureka2-server/src/main/java/com/netflix/eureka2/server/channel/InterestNotificationMultiplexer.java [hunk 5]
```diff
@@ -112,6 +112,6 @@ public void unregister() {
      * simplifies implementation of the multiplexer.
      */
     public Observable<ChangeNotification<InstanceInfo>> changeNotifications() {
-        return aggregatedStream;
+        return aggregatedStream.share();
```
Comment (tbak) [inline eureka2-server/src/main/java/com/netflix/eureka2/server/channel/InterestNotificationMultiplexer.java]: There is no need for that as channel contract guarantees single subscription only.

File: eureka2-client/src/main/java/com/netflix/eureka2/client/channel/InterestChannelInvoker.java (hunk 1 above)
Comment (qiangdavidliu) [inline eureka2-client/src/main/java/com/netflix/eureka2/client/channel/InterestChannelInvoker.java]: In actual implementation yes, but this is not enforced by the current InterestChannel contract. Going forward I propose we have all the channel interfaces extend the Sourced interface.

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java (hunk 2 above)
Comment (qiangdavidliu) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: Trigger happy autocomplete. Fixed.

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java (hunk 3 above)
Comment (qiangdavidliu) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: Legacy code not removed. Gone.

File: eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java (hunk 4 above)
Comment (qiangdavidliu) [inline eureka2-client/src/test/java/com/netflix/eureka2/client/interest/AbstractInterestClientTest.java]: In a previous iteration this was different. Removed.

File: eureka2-server/src/main/java/com/netflix/eureka2/server/channel/InterestNotificationMultiplexer.java (hunk 5 above)
Comment (qiangdavidliu) [inline eureka2-server/src/main/java/com/netflix/eureka2/server/channel/InterestNotificationMultiplexer.java]: As part of this pull InterestChannel have a new API getChangeNotificationStream(). On the server side this is bridged to the Multiplexer's changeNotifications so even though we don't subscribe to this stream more than once, this share() is to guard against that per the new contract.

**Pull request merged**